### ISSUE TITLE
[Request] pkg-config support

### ISSUE THREAD

**Issue by brown121407** [skypjack/entt#673]
_TL;DR: Please add pkg-config support to make it easier to depend on this library._

Hello,

I'm in the process of packaging EnTT for [GNU Guix](https://guix.gnu.org/), and I got a recipe that builds, passes the tests, and _even_ compiles projects when including the directory where the headers file end up! Here's [the code](https://paste.sr.ht/~brown121407/ea1fca164240c853e725a675dcc9b0a9bd94780e), if you're curious.

Now, that's great, but it means that users will need to add the include directory by hand in their projects. For example, to add EnTT to a Meson project I have to make my `meson.build` file look a bit like this:

```
...
incdir = include_directories('/home/user/.guix-profile/include/entt')
executable(..., include_directories: incdir)
```

(Maybe there are nicer ways to write that path, but you get the idea -- it has to be written manually.)

I don't know much about pkg-config, but I do know that when a project generates a `.pc` that ends up in the right place, build systems can read it and find the compiler options they need to use to include and link that library.

For example, the `fmt` library, which [builds a `.pc`](https://github.com/fmtlib/fmt/blob/6e1fc01752b3b4c0616d61ab642c714cc7a84670/CMakeLists.txt#L290) from [a template file](https://github.com/fmtlib/fmt/blob/master/support/cmake/fmt.pc.in) can be easily referenced by Meson: `dependency('fmt')` or even used in plain old Makefiles by embedding the pkg-config calls in them.

So my request is to that you would think about generating a `.pc` file, so we can include this library easier in our projects.

Thank you.

**Comment by skypjack** [comment]
Honestly, I know nothing about pkgconfig files, so any help is appreciated. 🙂 
Do you have any clue about how I can write something _that works_ ™️ ?

**Comment by brown121407** [comment]
Thanks for taking a look at this. Sadly, I haven't configured anything to work with pkgconfig until now, so I don't know what to advise. Knowing that you're open to this idea, if no one picks it up, I can try and come up with a PR at the end of this week.

**Comment by skypjack** [comment]
Well, if you are fine with contributing, it would be great. At least, it would provide a base from which to start for further refinements.
I'll try to look into how this kind of thing works too. Now I'm curious. 🙂

**Comment by skypjack** [comment]
[Yet another link](https://discord.com/channels/707607951396962417/707607952592470058/819223884334628864) to which to refer for refining the `pkgconfig` support. 🙂

**Comment by TerensTare** [comment]
@brown121407 Here is a minimal config I proposed. Suggestions are welcome

entt.pc.in
```pkgconfig
includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@

Name: entt
Description: Gaming meets modern C++ - a fast and reliable entity component system (ECS) and much more
Version: @entt_VERSION@
Cflags: -I${includedir}
```
and on CMakeLists.txt
```cpp
if(ENTT_ENABLE_INSTALL)
    # pkg-config support, except on Windows
    if(NOT WIN32 OR NOT CMAKE_HOST_SYSTEM_NAME MATCHES Windows)
        set(PKGCONFIG_INSTALL_DIR
            "${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig"
            CACHE PATH "Path where entt.pc is installed")

        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/entt.pc.in" "${CMAKE_CURRENT_BINARY_DIR}/entt.pc" @ONLY)
        install(FILES "${CMAKE_CURRENT_BINARY_DIR}/entt.pc" DESTINATION "${PKGCONFIG_INSTALL_DIR}")
    endif()
endif()
```

**Comment by brown121407** [comment]
@TerensTare, I am not an expert at CMake, so I want to check that I understand how to play with this. Does this mean that we would need to pass "-DENTT_ENABLE_INSTALL=ON" or something like that to the `cmake` command so it generates the `.pc` file?

**Comment by TerensTare** [comment]
Sol2 uses this method. However, given EnTT has no variable (like `ENTT_ENABLE_INSTALL`) for checking whether you want to "install" the project, and since pkgconfig doesn't work on Windows (according to the upmentioned code), I think we can safely remove the variable, and the code would become:
```cmake
    # pkg-config support, except on Windows
    if(NOT WIN32 OR NOT CMAKE_HOST_SYSTEM_NAME MATCHES Windows)
        set(PKGCONFIG_INSTALL_DIR
            "${CMAKE_INSTALL_DATAROOTDIR}/pkgconfig"
            CACHE PATH "Path where entt.pc is installed")

        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/entt.pc.in" "${CMAKE_CURRENT_BINARY_DIR}/entt.pc" @ONLY)
        install(FILES "${CMAKE_CURRENT_BINARY_DIR}/entt.pc" DESTINATION "${PKGCONFIG_INSTALL_DIR}")
    endif()
```

**Comment by skypjack** [comment]
> Sorry, that is just a Discord link on my side. Is there a possibility to share whatever you wanted without requiring the use of Discord?

😄 Sorry, my fault, I copied the wrong link.

**Comment by skypjack** [comment]
Please, can you confirm that branch `build_system` works as expected?
Also, I've reviewed a little the README file (same branch). Feel free to point out any error, integration or if you want to add a note about gnu guix under the integration section.
Thanks!

**Comment by brown121407** [comment]
Will do as soon as I can in one the following days and report back. Haven't got much free time lately...

**Comment by brown121407** [comment]
@skypjack, I can confirm that it works as expected, thank you. 

I will go ahead and send my recipes for `v3.6.0` to Guix. When the next release comes, if it includes the pkg-config stuff, after we'll update the package in Guix, I (or someone else) will probably come back and mention it in the README.

Please feel free to close this issue if you also think we're done here. Thank you for your collaboration!

**Comment by skypjack** [comment]
> I (or someone else) will probably come back and mention it in the README

Sure thing, I'd be glad to add it. I should cut a new release in a few days, so stay tuned. :slightly_smiling_face:

**Comment by skypjack** [comment]
Available on `master`. 👍
